Add ListenerSets documentation for NGF (#1925)
nginx/documentation: content/ngf/traffic-management/listener-sets.md

---

> title: ListenerSet API

Reviewer: I think we should have a more descriptive title. I'm not sure what that is at the moment, but I can think about it.

toc: true
weight: 1500
nd-content-type: how-to
nd-product: FABRIC

> nd-docs: DOCS-0000

Reviewer 2: lets add a description here. Check out other docs for examples. I also saw a PR for changing `nd-` to `f5-` 

Not sure if it applies here.

Author: can you expand on what you mean by description? Is that the docs number you're referring to? yea i'll change to the f5-

Reviewer 2: Sorry Ben, its not description. Its [summary](https://github.com/nginx/documentation/blob/main/content/ngf/traffic-security/oidc-authentication.md?plain=1#L8)

Linked an example. Also lets wait on the docs team to confirm the `f5-` change

Author: great, added a summary and description. feel free to review it or probably also the docs team will take a look

---

> Learn how to use the `ListenerSet` API.

Reviewer: This really doesn't add any value to the doc.

Author: How does "Learn how to use the `ListenerSet` API to specify additional listeners for a Gateway" sound?

Reviewer: That's better.

## Overview

A `ListenerSet` is a Gateway API type for specifying additional listeners for a Gateway. It decouples network listener configurations—such as ports, hostnames, and TLS termination—from the central Gateway resource and provides a mechanism to merge multiple listeners into a single Gateway. This enables multiple application developer teams who manage their own Services and Routes to configure their own listeners on a Gateway without needing to modify the Gateway itself (which may be owned by a different team).

> Additionally, by using ListenerSets, users are able to scale beyond the 64-listener limit of a single Gateway resource.

Reviewer 2: can we structure the benefits separately like the original guide ? i feel it looks structured

Author: could you elaborate more on what the original guide is? Are you referring to the gateway api docs?

Reviewer 2: Yeah sorry should have been clearer but you could list advantages like [this](https://gateway-api.sigs.k8s.io/guides/listener-set/) 

```
s such, you might use ListenerSets for the following advantages:

Multitenancy: You can let different teams create their own ListenerSets while sharing the same Gateway and backing load-balancing infrastructure.

Large scale deployments: By using ListenerSets, Gateways can have more than 64 listeners attached. Teams can also share the same ListenerSet configuration to avoid duplication.

Certificates for more listeners per gateway: Because you can now have more than 64 listeners per Gateway, a single Gateway can forward secured traffic to more backends that might have their own certificates. This approach aligns with projects that require service-level certificates, such as Istio Ambient Mesh or Knative.
``` 

Not specifically this but some advantages people get using it with NGF I guess

Author: I prefer the way it is now in more of a sentence description, i also believe it follows our other documents. i'll defer to docs team for this style/wording suggestion

Reviewer 3: We should not try to document 3rd party products, features, objects...

The guidance here is to keep it short and provide only what is strictly necessary, and link to the official K8s docs for more information in the subject.

Reviewer 2: The intent was not to document 3rd party product but to structure the important advantages here in bullet points. But I guess paragraph is fine so this should be good Ben

spec:
  replicas: 1
  selector:
    matchLabels:
      app: coffee
  template:
    metadata:
      labels:
        app: coffee
    spec:
      containers:
      - name: coffee
        image: nginxdemos/nginx-hello:plain-text
        ports:
        - containerPort: 8080
---
apiVersion: v1
kind: Service
metadata:
  name: coffee
spec:
  ports:
  - port: 80
    targetPort: 8080
    protocol: TCP
    name: http
  selector:
    app: coffee
EOF
```

This will create the `coffee` service and deployment. Run the following command to verify the resources were created:

```shell
kubectl get deployments,svc
```

Your output should include a `coffee` pod and service:

```text
NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/coffee   1/1     1            1           7s

NAME                 TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
service/coffee       ClusterIP   10.96.188.209   <none>        80/TCP    7s
```

Create a Gateway:

```yaml
kubectl apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: gateway
spec:
  gatewayClassName: nginx
  listeners:
    - name: http-example
      port: 80
      protocol: HTTP

>       hostname: "gateway.example.com"

Reviewer 2: is this listener never used? should we rename it like that?

Author: yea thats the listener thats not used, i don't think it deserves a specific naming difference.

Reviewer 2: but its confusing, just name it like `placeholder-listener` so it is clear what its use case is? I dont have a strong opinion on this but would be better in my approach

Author: adjusted to `gateway-listener` on the gateway, and listenerset-listener on the listenerset. I think thats a clearer naming convention than something related to it being unused

    Last Transition Time:  2026-05-04T23:36:50Z
    Message:               The Gateway is programmed
    Observed Generation:   1
    Reason:                Programmed
    Status:                True
    Type:                  Programmed
```

Save the public IP address and port(s) of the Gateway into shell variables:

```text
GW_IP=XXX.YYY.ZZZ.III
GW_PORT=<port number>
```

{{< call-out "note" >}}In a production environment, you should have a DNS record for the external IP address that is exposed, and it should refer to the hostname that the gateway will forward for.{{< /call-out >}}

## Configure a ListenerSet

For those that are familiar with the Gateway resource, ListenerSet `spec.listeners` is a direct copy of the Gateway's `spec.listeners`. For more information, view the [API Reference](https://gateway-api.sigs.k8s.io/reference/spec/#listenerset).

Create a ListenerSet:

```yaml
kubectl apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: ListenerSet
metadata:
  name: listenerset
spec:
  parentRef:
    name: gateway
    kind: Gateway
    group: gateway.networking.k8s.io
  listeners:
  - name: listenerset-http
    port: 80
    protocol: HTTP

>     hostname: "coffee.example.com"

Reviewer 2: either this needs updating or the curl command below that uses `cafe.example.com` or am I missing something?

Author: yea thanks for catching that, the curl command will change

      Status:                True
      Type:                  Programmed
      Last Transition Time:  2026-05-04T23:38:36Z
      Message:               The Listener is accepted
      Observed Generation:   1
      Reason:                Accepted
      Status:                True
      Type:                  Accepted
      Last Transition Time:  2026-05-04T23:38:36Z
      Message:               All references are resolved
      Observed Generation:   1
      Reason:                ResolvedRefs
      Status:                True
      Type:                  ResolvedRefs
      Last Transition Time:  2026-05-04T23:38:36Z
      Message:               No conflicts
      Observed Generation:   1
      Reason:                NoConflicts
      Status:                False
      Type:                  Conflicted
    Name:                    listenerset-http
    Supported Kinds:
      Group:  gateway.networking.k8s.io
      Kind:   HTTPRoute
      Group:  gateway.networking.k8s.io
      Kind:   GRPCRoute
```

The `Attached Routes` field should now report `1` as the HTTPRoute has attached to it successfully.

## Send traffic

Send a request to the `coffee` service with this command:

```shell

> curl --resolve cafe.example.com:$GW_PORT:$GW_IP http://cafe.example.com:$GW_PORT/coffee 

Reviewer 2: ```suggestion
curl --resolve coffee.example.com:$GW_PORT:$GW_IP http://coffee.example.com:$GW_PORT/coffee 
```

```

```text
Server address: 10.244.0.7:8080
Server name: coffee-654ddf664b-4xdrr
Date: 04/May/2026:23:45:57 +0000
URI: /coffee
Request ID: 7dbd29ec0c783475d50ed3b563b0a8a6
```

## See Also

> To set up HTTPS Termination or TLS passthrough on a listener from a `ListenerSet`, the configuration for the listener on the `ListenerSet` should be the same as it is on a Gateway. Follow our [HTTPS Termination]({{< ref "ngf/traffic-management/https-termination.md" >}}) and [TLS passthrough]({{<ref "ngf/traffic-management/tls-passthrough.md" >}}) guides and copy the Gateway listener's configuration onto a `ListenerSet` to mimic the behavior. 

Reviewer 2: ```suggestion
To set up HTTPS Termination or TLS passthrough on a listener from a `ListenerSet`, configure the listener on the `ListenerSet` exactly as you would on a Gateway. Follow our [HTTPS Termination]({{< ref "ngf/traffic-management/https-termination.md" >}}) and [TLS passthrough]({{<ref "ngf/traffic-management/tls-passthrough.md" >}}) guides and copy the Gateway listener's configuration onto a `ListenerSet` to mimic the behavior. 
```

Author: updated this, though left out the "exactly", felt a little too strong

---
title: ListenerSet API
toc: true
weight: 1500
f5-content-type: how-to
f5-product: FABRIC
f5-docs: DOCS-0000
f5-description: How to use the `ListenerSet` API to specify additional listeners for a Gateway.
f5-summary: >
    NGINX Gateway Fabric supports the ListenerSet Gateway API resource, which allows users to specify additional listeners for a Gateway.
    This reinforces Gateway API defined roles and personas by allowing application developers to configure their attachment points on a 
    Gateway without modifying the Gateway directly.

> ---

Reviewer 3: missing `f5-keywords`

    Last Transition Time:  2026-05-04T23:36:50Z
    Message:               The Gateway is programmed
    Observed Generation:   1
    Reason:                Programmed
    Status:                True
    Type:                  Programmed
```

Save the public IP address and port(s) of the Gateway into shell variables:

```text
GW_IP=XXX.YYY.ZZZ.III
GW_PORT=<port number>
```

{{< call-out "note" >}}In a production environment, you should have a DNS record for the external IP address that is exposed, and it should refer to the hostname that the gateway will forward for.{{< /call-out >}}

## Configure a ListenerSet

For those that are familiar with the Gateway resource, ListenerSet `spec.listeners` is a direct copy of the Gateway's `spec.listeners`. For more information, view the [API Reference](https://gateway-api.sigs.k8s.io/reference/spec/#listenerset).

Create a ListenerSet:

```yaml
kubectl apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: ListenerSet
metadata:
  name: listenerset
spec:
  parentRef:
    name: gateway
    kind: Gateway
    group: gateway.networking.k8s.io
  listeners:
  - name: listenerset-listener
    port: 80
    protocol: HTTP

>     hostname: "coffee.example.com"

Reviewer: For consistency with all of our other docs, can we keep this as `cafe.example.com`? Keeps the curl commands the same for all the docs.
